### ISSUE TITLE
Parse native threads from AppExitInfo for native crashes

### ISSUE THREAD

**Issue by romtsn** [getsentry/sentry-java#3295]
### Description

We can enrich native crashes even further by parsing [traceInputStream](https://developer.android.com/reference/android/app/ApplicationExitInfo#getTraceInputStream()) when the reason was [REASON_CRASH_NATIVE](https://developer.android.com/reference/android/app/ApplicationExitInfo#REASON_CRASH_NATIVE)

Bugsnag prior art https://github.com/bugsnag/bugsnag-android/tree/next/bugsnag-plugin-android-exitinfo/src/main/java/com/bugsnag/android

**Comment by supervacuus** [comment]
FYI: This is related with the investigation here: https://github.com/getsentry/sentry-native/issues/743#issuecomment-1279098537

**Comment by supervacuus** [comment]
However, tombstones won't necessarily decode what caused the `SIGABRT`. There is a good chance that the `SIGABRT` was an unhandled C++ exception, which would require a [terminate handler](https://en.cppreference.com/w/cpp/error/terminate) to decode what happened.

**Comment by Angelodaniel** [comment]
Hey Team, this came up again today for a customer. Any chance there is an update?

**Comment by OresteSalerno-TomTom** [comment]
Hi, we are encountering an issue on Android where plenty of different crashes are grouped together and only show `abort` in the stack trace. Is this improvement planned in a future release?

**Comment by supervacuus** [comment]
@OresteSalerno-TomTom and @Angelodaniel, this is now a high-priority topic for the Android SDK, and while there is no timeline yet, work on this basically starts now.

**Comment by supervacuus** [comment]
Quick update here: a first tombstone integration was released with [8.30.0](https://github.com/getsentry/sentry-java/releases/tag/8.30.0).

It is "minimal" in that it currently does not integrate with the Native SDK, so you need to enable either the NDK integration or the tombstone integration. If you enable both, you will get two events. Work on client-side deduplication (https://github.com/getsentry/sentry-java/pull/5037) between those two events is in progress.

Due to that limitation, the option to enable the tombstone integration is currently only exposed with "internal" interfaces. However, tombstone parsing and event creation are fully functional. They can already be tested in environments that don't rely on `sentry-native` context creation from native code and whose users run on Android 12+.

It's recommended that users default to [8.31.0](https://github.com/getsentry/sentry-java/releases/tag/8.31.0), as it downgrades `protobuf-lite` to `3.x`, which seems [less likely to conflict with transitive Firebase dependencies](https://github.com/getsentry/sentry-java/issues/5039). Options to drop the `protobuf-lite` dependency entirely are also evaluated.

**Comment by tustanivsky** [comment]
Some quick feedback from testing tombstone integration in Unreal SDK, specifically to verify whether thread information is attached to captured crash events on Android (related to https://github.com/getsentry/sentry-unreal/pull/1137).

As mentioned above, with NDK integration enabled two crash events are generated. The event captured via tombstone includes thread information, but its call stack does not appear to be relevant. Conversely, the event captured via the native integration has a correct call stack but no thread information:

- Crash captured with native integration ([example event](https://sentry-sdks.sentry.io/issues/6565469462/events/eb6b2bf911c046b77e512e065b70790e/)): correct call stack, but no thread info
- Crash captured with tombstone ([example event](https://sentry-sdks.sentry.io/issues/7229590204/?project=6253052&query=&referrer=issue-stream)): thread info present, but the call stack is misleading

Could this behavior be caused by having both crash capture mechanisms enabled at the same time? I’m currently using sentry-java `0.31.0`.

**Comment by supervacuus** [comment]
Hi @tustanivsky,

Thanks for this report! This appears to be a problem with the module extraction in the tombstone parser. It currently only retrieves executable memory maps, but this is not how Sentry or the Native SDK interprets the lossy module abstraction. It must include all readable memory maps for a given inode and assume the first memory map starts at file offset 0, so Sentry can correctly track runtime instruction addresses in the uploaded files.

I will add a fix for this to the deduplication PR (https://github.com/getsentry/sentry-java/pull/5037).

**Comment by tustanivsky** [comment]
Hey @supervacuus, just a quick follow-up on my previous attempts to use tombstones in the Unreal SDK (https://github.com/getsentry/sentry-unreal/pull/1137):

- Changes introduced in https://github.com/getsentry/sentry-java/pull/5037 resolved the issue with invalid crash call stacks captured by the tombstone integration and we are now receiving a single event with all required context and thread info.
- When using tombstones without the NDK enabled, captured crashes appear to be missing any trace details. This is likely not critical for Unreal as the `TombstoneMerged` mechanism will be used most of the time but it’s something to be aware of.
- When using tombstones, the `attachThreads` option set during Android SDK initialization is ignored and thread information is always attached regardless of its value. This isn’t a big deal either but it’s something I noticed during testing.

Here are also a few more example UE crash events for your reference:
- [NDK crash](https://sentry-sdks.sentry.io/issues/6565469462/events/eb6b2bf911c046b77e512e065b70790e/)
- [Tombstone-only crash](https://sentry-sdks.sentry.io/issues/6565469462/events/db1215a3d9b94b86822c03a7b2e9ca0b/)
- [Tombstone+NDK crash](https://sentry-sdks.sentry.io/issues/6565469462/events/17b6aa9e2c5c437e0dbef44f9b2423c4/)

**Comment by supervacuus** [comment]
> When using tombstones without the NDK enabled, captured crashes appear to be missing any trace details. This is likely not critical for Unreal as the TombstoneMerged mechanism will be used most of the time but it’s something to be aware of.

Yeah. The Tombstone integration is _**not a replacement**_ for the Native SDK for any direct or indirect users of the Native SDK, including all downstream SDKs that trace native code.

Once you disable the NDK integration, you will only get crash events with whatever enrichment `sentry-android` is able to recover. But no context created in the native layer will survive. As you said, for the Unreal SDK use case, both integrations should be enabled. 

In the best case, both can run, and Tombstones will improve your crash reports, and in the worst, you fall back to relying solely on the NDK integration. Using only the Tombstone integration can be useful for apps that use packages containing native code but do not instrument their own native code. I tried to touch on this in the docs: https://github.com/getsentry/sentry-docs/pull/16301, but I will likely stress this more.

> When using tombstones, the attachThreads option set during Android SDK initialization is ignored and thread information is always attached regardless of its value. This isn’t a big deal either but it’s something I noticed during testing.

That is a good point. I wasn't aware of this option. Will likely provide a fix for this or document why that is the case (seems like `AbnormalExit` implying `ANR` forces `attachThreads`). Will discuss this with the Android team.

**Comment by supervacuus** [comment]
@romtsn, since you created this issue, do you consider it resolved, or would you like to link it to the project closure?

**Comment by romtsn** [comment]
yeah, I'm closing it!
